### DISCUSSION TITLE
Elastix5.1.0 C++ project failure at elastix translation example

### DISCUSSION THREAD

**Discussion by Kunyadav** [SuperElastix/elastix#867] [General]
Hi All,
i have modified ElastixTranslationExample.cxx example by adding open cv api to read the moving and fixed image and below is my sample code. exceptions is getting thrown at line highlighted in bold below:
exception:Exception thrown at 0x00007FF606726D2D in elastix_translation_example.exe: 0xC0000005: Access violation reading location 0x0000000000000000.
Please advice and assist me in resolving this error.

#include <itkElastixRegistrationMethod.h>
#include <elxParameterObject.h>
#include <Core/elxVersionMacros.h>
#include <iostream>
#include <itkImage.h>
#include "itkImageFileReader.h"
#include "itkOpenCVImageBridge.h"
#include <itkSize.h>
#include <cassert>
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/videoio.hpp"
#include "opencv2/opencv.hpp"
#include "param.h"

using namespace std;

int main()
{
  enum
  {
    imageSizeX = 6,
    imageSizeY = 5
  };
   typedef itk::Image<unsigned char, 2> ImageType;
  cout << "Built with OpenCV 0" << CV_VERSION << endl;
  cv::Mat sem_input= cv::imread("C:/Users/e182000/Documents/cad2semm/waveform/B79Y22ADD/wf9_Meas.tif");  
  cv::Mat cad_input = cv::imread("C:/Users/e182000/Documents/cad2semm/waveform/B79Y22ADD/wf9_Model.tif");  
  ImageType::Pointer itkFixedImage = ImageType::New();  
  ImageType::Pointer itkMovingImage = ImageType::New();
 
  CADBasedCoarseEdgeRecipeParams *recipe_params = new CADBasedCoarseEdgeRecipeParams();
  recipe_params->setMaxNumOfIerationsFactor(0.1);
  cout << "Built with OpenCV 1-> " << CV_VERSION << endl;
  itkFixedImage = itk::OpenCVImageBridge::CVMatToITKImage< ImageType >(sem_input);
  cout << "Built with OpenCV 2-> " << CV_VERSION << endl;
  itkMovingImage = itk::OpenCVImageBridge::CVMatToITKImage< ImageType >(cad_input);
  cout << "Built with OpenCV 3->" << CV_VERSION << endl;
  cv::imshow("Sample", sem_input);
  cv::waitKey(0);
  cv::Mat image_blurred_with_3x3_kernel;
  cv::GaussianBlur(sem_input, image_blurred_with_3x3_kernel, cv::Size(3, 3), 0);
  cv::imshow("Blurred", image_blurred_with_3x3_kernel);
  cv::waitKey(0);
  const auto parameterObject = elx::ParameterObject::New();
  parameterObject->SetParameterMap(
    elx::ParameterObject::ParameterMapType{ { "ImageSampler", { "Full" } },
                                            { "MaximumNumberOfIterations", { "44" } },
                                            { "Metric", { "AdvancedNormalizedCorrelation" } },
                                            { "Optimizer", { "AdaptiveStochasticGradientDescent" } },
                                            { "Transform", { "TranslationTransform" } } });

  const auto filter = itk::ElastixRegistrationMethod<ImageType, ImageType>::New();
  filter->SetFixedImage(itkFixedImage);
  filter->SetMovingImage(itkMovingImage);
  filter->SetParameterObject(parameterObject);
  **filter->Update();**
 // filter->UpdateLargestPossibleRegion();

} // end main()

Regards
Kunal

**Comment by ntatsisk** [comment]
Hi @Kunyadav,
thanks for reposting here!

First things first, it is generally adviced when calling `Update()` on an ITK filter like `itk::ElastixRegistrationMethod` to do it inside a try-catch clause. So, in your example:

```
  try
    {
      filter->Update();
    }
  catch (itk::ExceptionObject & err)
    {
      std::cerr << "ExceptionObject caught !" << std::endl;
      std::cerr << err << std::endl;
      return EXIT_FAILURE;
    }
```
this might give a better indication of what the actual error is.

In addition to that, can you write to disk and inspect the `itkFixedImage` and `itkMovingImage`? Do they look as you expect or maybe they are empty?

**Comment by Kunyadav** [comment]
Hi, 
Thanks for the response.
i have added the try catch block but its not catching the exceptions. Also i have added null check for itkFixedImage  and itkMovingImage. though i did not write to the disk as i thought it is not required. but i am getting same exception at the same line filter->update();

#include <itkElastixRegistrationMethod.h>
#include <elxParameterObject.h>
#include <Core/elxVersionMacros.h>
#include <iostream>
#include <itkImage.h>
#include "itkImageFileReader.h"
#include "itkOpenCVImageBridge.h"
#include <itkSize.h>
#include <cassert>
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/videoio.hpp"
#include "opencv2/opencv.hpp"
#include "param.h"

using namespace std;

int
main()
{
  enum
  {
    imageSizeX = 6,
    imageSizeY = 5
  };
  try
  {
   typedef itk::Image<unsigned char, 2> ImageType;
  //typedef itk::Image<float> ImageType;
  cout << "Built with OpenCV 0" << CV_VERSION << endl;
  cv::Mat sem_input= cv::imread("C:/Users/e182000/Documents/cad2semm/waveform/B79Y22ADD/wf9_Meas.tif");  
  cv::Mat cad_input = cv::imread("C:/Users/e182000/Documents/cad2semm/waveform/B79Y22ADD/wf9_Model.tif");  
  ImageType::Pointer itkFixedImage = ImageType::New();  
  ImageType::Pointer itkMovingImage = ImageType::New();
 
  CADBasedCoarseEdgeRecipeParams *recipe_params = new CADBasedCoarseEdgeRecipeParams();
  recipe_params->setMaxNumOfIerationsFactor(0.1);
  cout << "Built with OpenCV 1-> " << CV_VERSION << endl;
  itkFixedImage = itk::OpenCVImageBridge::CVMatToITKImage< ImageType >(sem_input);
  cout << "Built with OpenCV 2-> " << CV_VERSION << endl;
  itkMovingImage = itk::OpenCVImageBridge::CVMatToITKImage< ImageType >(cad_input);
  cout << "Built with OpenCV 3->" << CV_VERSION << endl;
  cv::imshow("Sample", sem_input);
  cv::waitKey(0);
  cv::Mat image_blurred_with_3x3_kernel;
  cv::GaussianBlur(sem_input, image_blurred_with_3x3_kernel, cv::Size(3, 3), 0);
  cv::imshow("Blurred", image_blurred_with_3x3_kernel);
  cv::waitKey(0);

 if( itkFixedImage.IsNull()==true || itkMovingImage.IsNull()==true)
 {
   cout << "its Null" << endl;
   return EXIT_FAILURE;
 }
 else
 {
   cout << "Not Null\n";
 }  
  const auto parameterObject = elx::ParameterObject::New();
  parameterObject->SetParameterMap(
    elx::ParameterObject::ParameterMapType{ { "ImageSampler", { "Full" } },
                                            { "MaximumNumberOfIterations", { "44" } },
                                            { "Metric", { "AdvancedNormalizedCorrelation" } },
                                            { "Optimizer", { "AdaptiveStochasticGradientDescent" } },
                                            { "Transform", { "TranslationTransform" } } });

  const auto filter = itk::ElastixRegistrationMethod<ImageType, ImageType>::New();
  if (filter.IsNull() == true || parameterObject.IsNull()==true)
  {
    cout << "filter failed 4->" << CV_VERSION << endl;
    return EXIT_FAILURE;
  }
  else
  {
    cout << "all good till now" << endl;
  }
  filter->SetFixedImage(itkFixedImage);
  filter->SetMovingImage(itkMovingImage);
  filter->SetParameterObject(parameterObject);
 
    **filter->Update();**
    cout << "Update was successful" << endl;
    //filter->UpdateLargestPossibleRegion();
  }
  catch (itk::ExceptionObject & err)
  {
    std::cerr << "ExceptionObject caught !" << std::endl;
    std::cerr << err << std::endl;
    return EXIT_FAILURE;
  }
  catch (...)
  {
    std::cerr << "ExceptionObject caught !" << std::endl;
    return EXIT_FAILURE;
  }
 

} // end main()

Regards
Kunal

**Comment by ntatsisk** [comment]
I am afraid that checking for nulls is not sufficient in this case because that check just regards if the image pointers point to any location not if the underlying pixel array is correctly setup and populated.

Even if not needed for your code, you could do `itk::WriteImage(itkFixedImage, "fixed_image.tif")` etc. and then inspect the image in a viewer like ImageJ. Make sure to also include `#include <itkImageFileWriter.h>` on the top.

I think that ensuring that your images are converted correctly is the first step to debugging this code.

**Comment by Kunyadav** [comment]
Hi,
 i have tried saving **itkFixedImage and itkMovingImage** images to a tiff file like you have suggested, i can see images are getting saved correctly. still the filter->update() api is failing with "access violation" exceptions.
could you please help me resolving this issue? i am really new to elastix uses.

#include <itkElastixRegistrationMethod.h>
#include <elxParameterObject.h>
#include <Core/elxVersionMacros.h>
#include <iostream>
#include <itkImage.h>
#include "itkImageFileReader.h"
#include "itkOpenCVImageBridge.h"
#include <itkSize.h>
#include <cassert>
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/videoio.hpp"
#include "opencv2/opencv.hpp"
#include "param.h"
#include "itkImage.h"
#include "itkImageFileWriter.h"
using namespace std;

int
main()
{
  enum
  {
    imageSizeX = 6,
    imageSizeY = 5
  };
  try
  {
   typedef itk::Image<unsigned char, 2> ImageType;
  //typedef itk::Image<float> ImageType;
  cout << "Built with OpenCV 0" << CV_VERSION << endl;
  cv::Mat sem_input= cv::imread("C:/Users/e182000/Documents/cad2semm/waveform/B79Y22ADD/wf9_Meas.tif");  
  cv::Mat cad_input = cv::imread("C:/Users/e182000/Documents/cad2semm/waveform/B79Y22ADD/wf9_Model.tif");  
  ImageType::Pointer itkFixedImage = ImageType::New();  
  ImageType::Pointer itkMovingImage = ImageType::New();
 
  CADBasedCoarseEdgeRecipeParams *recipe_params = new CADBasedCoarseEdgeRecipeParams();
  recipe_params->setMaxNumOfIerationsFactor(0.1);
  itkFixedImage = itk::OpenCVImageBridge::CVMatToITKImage< ImageType >(sem_input);
  itkMovingImage = itk::OpenCVImageBridge::CVMatToITKImage< ImageType >(cad_input);

  using WriterType = itk::ImageFileWriter<ImageType>;
  auto writer = WriterType::New();
  auto writerImg = WriterType::New();;
  writerImg->SetFileName("F:\\elast\\Moving_image.tif");
  writerImg->SetInput(itkMovingImage);

  writer->SetFileName("F:\\elast\\fixed_image.tif");
  writer->SetInput(itkFixedImage);
  try
  {
    writer->Update();
    writerImg->Update();
  }
  catch (const itk::ExceptionObject & error)
  {
    std::cerr << "Error: " << error << std::endl;
    return EXIT_FAILURE;
  }
  cv::imshow("Sample", sem_input);
  cv::waitKey(0);
  cv::Mat image_blurred_with_3x3_kernel;
  cv::GaussianBlur(sem_input, image_blurred_with_3x3_kernel, cv::Size(3, 3), 0);
  cv::imshow("Blurred", image_blurred_with_3x3_kernel);
  cv::waitKey(0);

 if( itkFixedImage.IsNull()==true || itkMovingImage.IsNull()==true)
 {
   cout << "its Null" << endl;
   return EXIT_FAILURE;
 }
 else
 {
   cout << "Not Null\n";
 }  
  const auto parameterObject = elx::ParameterObject::New();
  parameterObject->SetParameterMap(
    elx::ParameterObject::ParameterMapType{ { "ImageSampler", { "Full" } },
                                            { "MaximumNumberOfIterations", { "44" } },
                                            { "Metric", { "AdvancedNormalizedCorrelation" } },
                                            { "Optimizer", { "AdaptiveStochasticGradientDescent" } },
                                            { "Transform", { "TranslationTransform" } } });

  const auto filter = itk::ElastixRegistrationMethod<ImageType, ImageType>::New();
  if (filter.IsNull() == true || parameterObject.IsNull()==true)
  {
    cout << "filter failed 4->" << CV_VERSION << endl;
    return EXIT_FAILURE;
  }
  else
  {
    cout << "all good till now" << endl;
  }
  filter->SetFixedImage(itkFixedImage);
  filter->SetMovingImage(itkMovingImage);
  filter->SetParameterObject(parameterObject);
 
    **filter->Update();**
    cout << "Update was successful" << endl;
  }
  catch (itk::ExceptionObject & err)
  {
    std::cerr << "ExceptionObject caught !" << std::endl;
    std::cerr << err << std::endl;
    return EXIT_FAILURE;
  }
  catch (...)
  {
    std::cerr << "ExceptionObject caught !" << std::endl;
    return EXIT_FAILURE;
  }
 

} // end main()

**Comment by Kunyadav** [comment]
Hi,
I have further debugged and found it is crashing at below line in the image:


![image](https://user-images.githubusercontent.com/124020688/230852645-e753d095-d722-4650-91ac-483f620b78e8.png)

Any help will be really appreciable. 
Regards
Kunal

**Reply by N-Dekker** [reply]
Hi Kunal, your screenshot might be helpful  😃  It appears that the function call at line 49 of elxResamplerBase.hxx has returned null, right?

    FixedImageType * fixedImage = this->m_Elastix->GetFixedImage();

Could you possibly put a breakpoint inside `ElastixTemplate::GetFixedImage(unsigned)` and then rerun your example?

https://github.com/SuperElastix/elastix/blob/c3d478e8e3b6b1806862abbd0d777beff9b66346/Core/Kernel/elxElastixTemplate.hxx#L86

I suspect that the `dynamic_cast<FixedImageType *>` returns nullptr, in your case. If it does, please check the actual type of the image stored in the container returned by GetFixedImageContainer(), and the expected image type `FixedImageType`.  Are they the same, or different?

**Comment by Kunyadav** [comment]
Hi N-Dekker,
 you are right, while debugging i have found  there is a difference between the image type . FixedImageType is of itk::Image<float,2> but GetFixedImageContainer() is of itk::Image<unsigned char,2> . my code has defined "typedef itk::Image<unsigned char, 2> ImageType"
Why the FixedImageType is expecting of itk::Image<float,2> type? how can i change it to itk::Image<unsigned char,2>?
Regards
Kunal

**Reply by N-Dekker** [reply]
Glad you found it! 👍 One way to solve it would be to reconfigure elastix in CMake by adding `unsigned char` to the `ELASTIX_IMAGE_2D_PIXELTYPES` CMake variable.  (Separated by a semi-colon, for example `unsigned char;float`) But this would require a complete rebuild of your elastix library. Maybe it can be fixed in a more convenient way. I can have a closer look later.

PS I think you should also explicitly specify the pixel types with the parameter map, for example:

    { "FixedInternalImagePixelType", { "unsigned char" } },
    { "MovingInternalImagePixelType", { "unsigned char" } },

**Comment by kyadav78** [comment]
Hi N-Dekker,
 did you get a chance to look into this issue? is there any convenient way to make it work for pixel type unsigned char? 
when i am using pixelt type as "float" type my code was not working .
Below code does not give proper results.
 filter->UpdateLargestPossibleRegion();



Regards
Kunal

**Reply by N-Dekker** [reply]
Hi Kunal,
We looked at your issue again, and agreed that pixel type support could indeed be made more convenient  😃 Specifically, this pull request should address the inconvenience:
 - #882 (Convert the input images to the user-specified internal pixel type)

Which was merged to the `main` branch of elastix on May 8: commit 58e0a7bdd76f69a8615cb4a8856038c0892ea530

Can you please try it out (you may just rebuild the latest revision from its `main` branch), and tell us if it does indeed work for you?

Regards, Niels

**Comment by kyadav78** [comment]
Hi N-Dekker,
 Thank you for your support till now.
I have raised another issue https://github.com/SuperElastix/elastix/issues/932 but i am not getting any response there, so i am asking it here, when i am running above code for fixed and moving image, it is taking more than 30 seconds to complete the call .
filter->update() or filter->UpdateLargestPossibleRegion() API. is there any way to reduce the execution time for this api?
Regards
Kunal
